File: public/.julia/v0.5/mosek/deps/src/mosek/8/tools/examples/fusion/python/lpt.py
##
#    Copyright: Copyright (c) MOSEK ApS, Denmark. All rights reserved.
#
#    File:    lpt.py
#
#    Purpose:  Demonstrates how to solve the multi-processor
#              scheduling problem using the Fusion API.
##


import sys
import random
import mosek
from mosek.fusion import *


def main():
    #Parameters:
    n= 1000             #Number of tasks
    m= 8                #Number of processors

    lb= 1.              #Range of short task lengths
    ub= 5.

    sh= 0.8             #Proportion of short tasks
    n_short= int(n*sh)
    n_long= n-n_short

    random.seed(0)
    T= sorted( [random.uniform(lb,ub) for i in range(n_short)]
              +[random.uniform(20*lb,20*ub) for i in range(n_long)], reverse=True)

    print("# jobs(n)   : ",n)
    print("# machine(m): ",m)

    ##TAG:begin-model
    with Model('Multi-processor scheduling') as M:

        x= M.variable('x', [m,n], Domain.binary())

        t= M.variable('t',1, Domain.unbounded())

        M.constraint( Expr.sum(x,0), Domain.equalsTo(1.) )

    ##TAG:begin-repeat
        M.constraint( Expr.sub( Var.repeat(t,m), Expr.mul(x,T) ) , Domain.greaterThan(0.) )
    ##TAG:end-repeat

        M.objective( ObjectiveSense.Minimize, t )
    ##TAG:end-model

    ##TAG:begin-lpt
        #LPT heuristic
        schedule=[0. for i in range(m)]
        init= [0. for i in range(n*m)]

        for i in range(n):
            mm= schedule.index(min(schedule))
            schedule[mm]+= T[i]
            init[n*mm+i]=1.
    ##TAG:end-lpt
      
        #Comment this line to switch off feeding in the initial LPT solution  
    ##TAG:begin-sol-input    
        x.setLevel(init)
    ##TAG:end-sol-input

        M.setLogHandler(sys.stdout)  
        M.setSolverParam("mioTolRelGap", .01)
        M.solve()

        print('initial solution:')
        for i in range(m):
            print('M',i,init[i*n:(i+1)*n])

        print('MOSEK solution:')
        for i in range(m):
            print('M',i,[y for y in x.slice([i,0],[i+1,n]).level()])
    
if __name__ == '__main__':
    main()

File: public/.julia/v0.5/mosek/deps/src/mosek/8/tools/examples/fusion/dotnet/lpt.cs
//
// Copyright: Copyright (c) MOSEK ApS, Denmark. All rights reserved.
//
//  File:      lpt.cs
//
// Purpose: demonstrates how to feed the MIP solver with some initial solution.
//

using System;
using System.Collections.Generic;

using mosek.fusion;


namespace mosek
{
  namespace fusion
  {
    namespace example
    {

        public class lpt
        {

            public static void Main(string [] args)
            {
              
              double lb = 1.0;
              int m = 4;
              int n= 15;
              int seed= 0;
              double ub = 5.0;
              
              double[] T= new double[n]; 

              Random gen = new Random(seed);
              for(int i=0;i<n;i++)
                T[i] = gen.NextDouble()*(ub-lb)+lb;

              Array.Sort<double>(T); 
      
              //TAG:begin-model
              Model M= new Model("Multi-processor scheduling");

              Variable x= M.Variable("x", new int[]{m, n}, Domain.Binary());

              Variable t= M.Variable("t",1, Domain.Unbounded());

              M.Constraint( Expr.Sum(x,0), Domain.EqualsTo(1.0));

              //TAG:begin-repeat
              M.Constraint( Expr.Sub( Var.Repeat(t,m), Expr.Mul(x,T)) , Domain.GreaterThan(0.0));
              //TAG:end-repeat

              M.Objective( ObjectiveSense.Minimize, t );
              //TAG:end-model

              //TAG:begin-lpt
              double [] init    = new double[n*m];
              double [] schedule= new double[m];
        
        
              for(int i=0;i<n;i++)
              {
                int next=0;
                for(int j=1;j<m;j++)
                  if(schedule[j]<schedule[next]) next=j;
          
                  schedule[next] += T[i];
                  init[next*n + i] = 1;
              }
              //TAG:end-lpt
      
              M.SetLogHandler(Console.Out);

              M.WriteTask("dump.opf");
                    
              M.SetSolverParam("mioTolRelGap", .01);
              M.Solve();

              try
              {
                  Console.WriteLine("\n\nInitial solution: ");
                  for (int i = 0; i < m; i++)
                  {
                      Console.Write("M {0} [", i);
                      for (int y = 0; y < n; y++)
                          Console.Write(init[i * n + y] + ", ");
                      Console.WriteLine("]");
                  }
                  //TAG:begin-sol-input    
                  x.SetLevel(init);
                  //TAG:end-sol-input    


                 Console.WriteLine("\n\nMOSEK solution:");
                 for (int i = 0; i < m; i++)
                 {
                   Console.Write("M {0} [",i);
                   for(int y=0; y<n;y++)
                   {
                      double value = x.Index(i,y).Level()[0];
                       Console.Write( "{0}, ", (int)value);
                   }
                   Console.WriteLine("]");
                 }
              }
              catch(SolutionError e)
              {
                //must do something about....
              }
            }
      
        }

    }
  }
}
